Feature/phantom screenshots

elessar-ch opened the pull request:
This PR includes everything needed to create and compare screenshots during the build.
The process is executed in a secondary rancher-container.

elessar-ch committed:
Add first draft of making screenshots

elessar-ch committed:
Make sure all pages are rendered?

elessar-ch committed:
Load stylesheet for demos in the head

elessar-ch committed:
Do screenshots in multiple viewports

elessar-ch committed:
Draft changes for comparing screenshots

elessar-ch committed:
Refactor the screenshot scripts

elessar-ch committed:
Output diff image after comparing screenshots

elessar-ch committed:
Add package.json and Dockerize the whole thing

elessar-ch committed:
Save screenshots in specified path

elessar-ch committed:
Compare all the screenshots recorded

elessar-ch committed:
Add fonts to phantomjs image.

elessar-ch committed:
Add diff directory in docker image

elessar-ch committed:
Add font-test demo page

elessar-ch committed:
Fix test definitions

elessar-ch committed:
Add test for fonts

elessar-ch committed:
Fix expected pages

elessar-ch committed:
Fix expected file for font-test

elessar-ch committed:
Try adding docker image for tests to jenkinsfile

elessar-ch committed:
Fix typo in path in jenkinsfile

elessar-ch committed:
Start the test container along the web container

elessar-ch committed:
Add link to web container

elessar-ch committed:
Make run-tests executable

elessar-ch committed:
Start test service only once

elessar-ch committed:
Add test image parameter to rancher compose commands

elessar-ch committed:
Change screenshot paths

elessar-ch committed:
Add results page

elessar-ch committed:
Add handlebars dependency for results page

elessar-ch committed:
Integrate test results into toolkit

elessar-ch committed:
Name link in test container

elessar-ch committed:
Reuse volume from the toolkit container for test

elessar-ch committed:
Specify test container as sidekick to share volume

elessar-ch committed:
Use branch-specific BASE_URL

elessar-ch committed:
Add readme to tests

davidknezic commented on Jenkinsfile:
```diff
@@ -5,8 +5,9 @@ node {
+    def testImageName = repositoryName + "-test" + ":" + tagName.replaceAll("[.:_/\\\\]", '-')
     def projectName = imageName.replaceAll("[.:_/\\\\]", '-')
-    def baseUrl = '/toolkit'
+    def baseUrl = '/' + tagName.replaceAll("[.:_/\\\\]", '-')
```
Isn't this going to break the `master` build, that would now have a baseUrl `/latest` instead of `/toolkit`?

davidknezic commented on tests/compare.js:
```diff
@@ -0,0 +1,41 @@
+
+console.log('All tests finished. Writing results page.')
+
+resultPage.createResultPage(pages)
```
Could you rewrite this to ES6?
Also make sure the `.eslintrc.json` file is correctly loaded in your editor. 
Making sure your editor reads the `.editorconfig` file will prevent those missing end lines in all your files.

davidknezic commented on tests/package.json:
```diff
@@ -0,0 +1,37 @@
+    "css",
+    "js"
+  ],
+  "main": "test2.js",
```
That file doesn't exist.

davidknezic commented on tests/readme.md:
```diff
@@ -0,0 +1,25 @@
+# Web Toolkit Testing
+
+This is a docker image containing phantomjs and resemblejs to capture
+screenshots and compare them with expected values respecitely.
```
> respecitely

davidknezic commented on tests/template.hbs:
```diff
@@ -0,0 +1,41 @@
+        {{#each tests}}
+          <h2>{{name}}</h2>
+          {{#gt misMatchPercentage 0}}
+            u fucked up!
```
Watch out for wording here.

davidknezic reviewed (changes requested):
See comments. Also, we don't need all those fonts, only Source Sans Pro 400, 600, 700. Make sure your editors reads the `.editorconfig` file properly, because there are several formatting problems.

elessar-ch committed:
Merge branch 'master' into feature/phantom-screenshots

Conflicts:
	.dockerignore
	.npmignore
	Dockerfile
	docs/hbs-partials/demo.hbs

elessar-ch committed:
Extend the readme about tests with phantom a bit

elessar-ch committed:
Remove PT Serif from testing folder

elessar-ch committed:
Remove console logs from test

elessar-ch committed:
Update test definitions

elessar-ch committed:
Update tests/package.json

elessar-ch committed:
Fix typo in test/readme

elessar-ch committed:
Update jenkinsfile to keep the /toolkit base_url on master

elessar-ch committed:
Update test result page

elessar-ch commented:
Currently working on making this thing a bit cooler :P

elessar-ch committed:
Use ES6 to take snapshots for testing

elessar-ch committed:
Do compare with ES6

elessar-ch committed:
Only notify about errors when they actually happened

elessar-ch committed:
Await execution of tests before writing resultpage

elessar-ch committed:
Add info about testing locally to test/readme.md

elessar-ch committed:
Use ES6 in result-page

elessar-ch committed:
Fix usage of result-page

elessar-ch committed:
Add passed/failed message to top of result-page

elessar-ch committed:
Mention -d flag in test/readme.md

elessar-ch committed:
Use BASE_URL env var again when doing screenshots

elessar-ch commented:
Allright, I think we're close to merging now.
- Moved to ES6
- Doing the snapshots now in node too (with PhantomJS Node bridge)
- Improved result page with some colors (don't get too excited)
- Removed PT Serif
- Fixed typos in readme :P and extended it a bit

davidknezic reviewed (changes requested):
Just three things
- `docker-compose.yml` and `font-test.hbs` still have no newlines
  
  ![screen shot 2016-11-11 at 11 30 01](https://cloud.githubusercontent.com/assets/198988/20212231/4d94266a-a802-11e6-9919-a24476d640f8.png)
- You can remove three more fonts, because we don't use italics anywhere (and hopefully never will)
  - SourceSansPro-BoldIt.ttf
  - SourceSansPro-It.ttf
  - SourceSansPro-SemiboldIt.ttf
- And I'd prefer if the `tests/readme.md` filename was capsed `README.md`.

Everything else, amazing! 🎉

elessar-ch committed:
Fix minor issues regarding tests

elessar-ch committed:
Rename readme.md to README.md

elessar-ch commented:
@davidknezic fixed

davidknezic reviewed (approved):
👍
